File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    dy = [0,1,0,-1]
    dx = [1,0,-1,0]
    y,x = 0, -1
    
    arr = [[0] * n for _ in range(n)]
    cnt = 1
    
    dir = 0
    
    while cnt <= n**2:
        ny, nx = y+dy[dir], x+dx[dir]
        if 0<=ny < n and 0<=nx < n and not arr[ny][nx]:
            arr[ny][nx] = cnt
            cnt += 1
            y, x = ny, nx
        else:
            dir = (dir+1) % 4
    
    return arr

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.java
class Solution {
    public int[][] solution(int n) {
        
        if(n == 1) return new int[][]{{1}};
        
        int[][] answer = new int[n][n];
        
        int cnt = 1;
        
        for(int itr =0; itr < n/2; itr++){
            
            int startX = itr;
            int startY = itr;
                        
            for(; startY < (n-1)-itr; startY++){
                answer[startX][startY] = cnt++;
            }
            
            for(; startX < (n-1)-itr; startX++){
                answer[startX][startY] = cnt++;
            }
            
            for(; startY >itr; startY--){
                answer[startX][startY] = cnt++;
            }
            
            for(; startX > itr; startX--){
                answer[startX][startY] = cnt++;
            }
        }
        
        if(n%2 != 0){
            answer[n/2][n/2] = cnt;
        }
        
        
        return answer;
    }
}